File: fontes/backend/delphi/dmvc/Service/Tributacao/TributGrupoTributarioService.pas
{*******************************************************************************
Title: T2Ti ERP Fenix                                                                
Description: Service relacionado à tabela [TRIBUT_GRUPO_TRIBUTARIO] 
                                                                                
The MIT License                                                                 
                                                                                
Copyright: Copyright (C) 2020 T2Ti.COM                                          
                                                                                
Permission is hereby granted, free of charge, to any person                     
obtaining a copy of this software and associated documentation                  
files (the "Software"), to deal in the Software without                         
restriction, including without limitation the rights to use,                    
copy, modify, merge, publish, distribute, sublicense, and/or sell               
copies of the Software, and to permit persons to whom the                       
Software is furnished to do so, subject to the following                        
conditions:                                                                     
                                                                                
The above copyright notice and this permission notice shall be                  
included in all copies or substantial portions of the Software.                 
                                                                                
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,                 
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES                 
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND                        
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT                     
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,                    
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING                    
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR                   
OTHER DEALINGS IN THE SOFTWARE.                                                 
                                                                                
       The author may be contacted at:                                          
           t2ti.com@gmail.com                                                   
                                                                                
@author Albert Eije (alberteije@gmail.com)                    
@version 1.0.0
*******************************************************************************}
unit TributGrupoTributarioService;

interface

uses
  TributGrupoTributario, 
  System.SysUtils, System.Generics.Collections, ServiceBase, MVCFramework.DataSet.Utils;

type

  TTributGrupoTributarioService = class(TServiceBase)
  private
    class procedure AnexarObjetosVinculados(AListaTributGrupoTributario: TObjectList<TTributGrupoTributario>); overload;
    class procedure AnexarObjetosVinculados(ATributGrupoTributario: TTributGrupoTributario); overload;
  public
    class function ConsultarLista: TObjectList<TTributGrupoTributario>;
    class function ConsultarListaFiltro(AWhere: string): TObjectList<TTributGrupoTributario>;
    class function ConsultarObjeto(AId: Integer): TTributGrupoTributario;
    class procedure Inserir(ATributGrupoTributario: TTributGrupoTributario);
    class function Alterar(ATributGrupoTributario: TTributGrupoTributario): Integer;
    class function Excluir(ATributGrupoTributario: TTributGrupoTributario): Integer;
  end;

var
  sql: string;


implementation

{ TTributGrupoTributarioService }

class procedure TTributGrupoTributarioService.AnexarObjetosVinculados(ATributGrupoTributario: TTributGrupoTributario);
begin
end;

class procedure TTributGrupoTributarioService.AnexarObjetosVinculados(AListaTributGrupoTributario: TObjectList<TTributGrupoTributario>);
var
  TributGrupoTributario: TTributGrupoTributario;
begin
  for TributGrupoTributario in AListaTributGrupoTributario do
  begin
    AnexarObjetosVinculados(TributGrupoTributario);
  end;
end;

class function TTributGrupoTributarioService.ConsultarLista: TObjectList<TTributGrupoTributario>;
begin
  sql := 'SELECT * FROM TRIBUT_GRUPO_TRIBUTARIO ORDER BY ID';
  try
    Result := GetQuery(sql).AsObjectList<TTributGrupoTributario>;
    AnexarObjetosVinculados(Result);
  finally
    Query.Close;
    Query.Free;
  end;
end;

class function TTributGrupoTributarioService.ConsultarListaFiltro(AWhere: string): TObjectList<TTributGrupoTributario>;
begin
  sql := 'SELECT * FROM TRIBUT_GRUPO_TRIBUTARIO where ' + AWhere;
  try
    Result := GetQuery(sql).AsObjectList<TTributGrupoTributario>;
    AnexarObjetosVinculados(Result);
  finally
    Query.Close;
    Query.Free;
  end;
end;

class function TTributGrupoTributarioService.ConsultarObjeto(AId: Integer): TTributGrupoTributario;
begin
  sql := 'SELECT * FROM TRIBUT_GRUPO_TRIBUTARIO WHERE ID = ' + IntToStr(AId);
  try
    GetQuery(sql);
    if not Query.Eof then
    begin
      Result := Query.AsObject<TTributGrupoTributario>;
      AnexarObjetosVinculados(Result);
    end
    else
      Result := nil;
  finally
    Query.Close;
    Query.Free;
  end;
end;

class procedure TTributGrupoTributarioService.Inserir(ATributGrupoTributario: TTributGrupoTributario);
begin
  ATributGrupoTributario.ValidarInsercao;
  ATributGrupoTributario.Id := InserirBase(ATributGrupoTributario, 'TRIBUT_GRUPO_TRIBUTARIO');
  
end;

class function TTributGrupoTributarioService.Alterar(ATributGrupoTributario: TTributGrupoTributario): Integer;
begin
  ATributGrupoTributario.ValidarAlteracao;
  Result := AlterarBase(ATributGrupoTributario, 'TRIBUT_GRUPO_TRIBUTARIO');
  
  
end;


class function TTributGrupoTributarioService.Excluir(ATributGrupoTributario: TTributGrupoTributario): Integer;
begin
  ATributGrupoTributario.ValidarExclusao;
  
  Result := ExcluirBase(ATributGrupoTributario.Id, 'TRIBUT_GRUPO_TRIBUTARIO');
end;


end.

File: fontes/frontend/fenix_provider/lib/src/service/tributacao/tribut_grupo_tributario_service.dart
/*
Title: T2Ti ERP 3.0                                                                
Description: Service relacionado à tabela [TRIBUT_GRUPO_TRIBUTARIO] 
                                                                                
The MIT License                                                                 
                                                                                
Copyright: Copyright (C) 2021 T2Ti.COM                                          
                                                                                
Permission is hereby granted, free of charge, to any person                     
obtaining a copy of this software and associated documentation                  
files (the "Software"), to deal in the Software without                         
restriction, including without limitation the rights to use,                    
copy, modify, merge, publish, distribute, sublicense, and/or sell               
copies of the Software, and to permit persons to whom the                       
Software is furnished to do so, subject to the following                        
conditions:                                                                     
                                                                                
The above copyright notice and this permission notice shall be                  
included in all copies or substantial portions of the Software.                 
                                                                                
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,                 
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES                 
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND                        
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT                     
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,                    
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING                    
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR                   
OTHER DEALINGS IN THE SOFTWARE.                                                 
                                                                                
       The author may be contacted at:                                          
           t2ti.com@gmail.com                                                   
                                                                                
@author Albert Eije (alberteije@gmail.com)                    
@version 1.0.0
*******************************************************************************/
import 'dart:convert';
import 'dart:typed_data';

import 'package:http/http.dart' show Client;

import 'package:fenix/src/service/service_base.dart';
import 'package:fenix/src/model/filtro.dart';
import 'package:fenix/src/model/model.dart';

/// classe responsável por requisições ao servidor REST
class TributGrupoTributarioService extends ServiceBase {
  var clienteHTTP = Client();

  Future<List<TributGrupoTributario>?> consultarLista({Filtro? filtro}) async {
    List<TributGrupoTributario> listaTributGrupoTributario = [];

    try {
      tratarFiltro(filtro, '/tribut-grupo-tributario/');
	  	
      final response = await clienteHTTP.get(Uri.tryParse(url)!, headers: ServiceBase.cabecalhoRequisicao);

      if (response.statusCode == 200) {
        if (response.headers["content-type"]!.contains("html")) {
          tratarRetornoErro(response.body, response.headers);
          return null;
        } else {
          var parsed = json.decode(response.body) as List<dynamic>;
          for (var tributGrupoTributario in parsed) {
            listaTributGrupoTributario.add(TributGrupoTributario.fromJson(tributGrupoTributario));
          }
          return listaTributGrupoTributario;
        }
      } else {
        tratarRetornoErro(response.body, response.headers);
        return null;
      }
    } on Exception catch (e) {
      tratarRetornoErro(null, null, exception: e);
    }
    return null;
  }

  Future<TributGrupoTributario?> consultarObjeto(int id) async {
    try {
      final response = await clienteHTTP.get(Uri.tryParse('$endpoint/tribut-grupo-tributario/$id')!, headers: ServiceBase.cabecalhoRequisicao);

      if (response.statusCode == 200) {
        if (response.headers["content-type"]!.contains("html")) {
          tratarRetornoErro(response.body, response.headers);
          return null;
        } else {
          var tributGrupoTributarioJson = json.decode(response.body);
          return TributGrupoTributario.fromJson(tributGrupoTributarioJson);
        }
      } else {
        tratarRetornoErro(response.body, response.headers);
        return null;
      }
    } on Exception catch (e) {
      tratarRetornoErro(null, null, exception: e);
    }
    return null;
  }

  Future<TributGrupoTributario?> inserir(TributGrupoTributario tributGrupoTributario) async {
    try {
      final response = await clienteHTTP.post(
        Uri.tryParse('$endpoint/tribut-grupo-tributario')!,
        headers: ServiceBase.cabecalhoRequisicao,
        body: tributGrupoTributario.objetoEncodeJson(tributGrupoTributario),
      );

      if (response.statusCode == 200 || response.statusCode == 201) {
        if (response.headers["content-type"]!.contains("html")) {
          tratarRetornoErro(response.body, response.headers);
          return null;
        } else {
          var tributGrupoTributarioJson = json.decode(response.body);
          return TributGrupoTributario.fromJson(tributGrupoTributarioJson);
        }
      } else {
        tratarRetornoErro(response.body, response.headers);
        return null;
      }
    } on Exception catch (e) {
      tratarRetornoErro(null, null, exception: e);
    }
    return null;
  }

  Future<TributGrupoTributario?> alterar(TributGrupoTributario tributGrupoTributario) async {
    try {
      var id = tributGrupoTributario.id;
      final response = await clienteHTTP.put(
        Uri.tryParse('$endpoint/tribut-grupo-tributario/$id')!,
        headers: ServiceBase.cabecalhoRequisicao,
        body: tributGrupoTributario.objetoEncodeJson(tributGrupoTributario),
      );

      if (response.statusCode == 200) {
        if (response.headers["content-type"]!.contains("html")) {
          tratarRetornoErro(response.body, response.headers);
          return null;
        } else {
          var tributGrupoTributarioJson = json.decode(response.body);
          return TributGrupoTributario.fromJson(tributGrupoTributarioJson);
        }
      } else {
        tratarRetornoErro(response.body, response.headers);
        return null;
      }
    } on Exception catch (e) {
      tratarRetornoErro(null, null, exception: e);
    }
    return null;
  }

  Future<bool?> excluir(int id) async {
    try {
      final response = await clienteHTTP.delete(
        Uri.tryParse('$endpoint/tribut-grupo-tributario/$id')!,
        headers: ServiceBase.cabecalhoRequisicao,
      );

      if (response.statusCode == 200) {
        return true;
      } else {
        tratarRetornoErro(response.body, response.headers);
        return null;
      }
    } on Exception catch (e) {
      tratarRetornoErro(null, null, exception: e);
    }
    return null;
  }
  
  Future<Uint8List?> visualizarPdf({Filtro? filtro}) async {
    try {
      tratarFiltroRelatorio(filtro, 'TributGrupoTributario');
      final response = await clienteHTTP.get(Uri.tryParse(urlRelatorios)!);
      if (response.statusCode == 200) {
        if (response.headers["content-type"]!.contains("html")) {
          tratarRetornoErro(response.body, response.headers);
          return null;
        } else {
          final pdf = response.bodyBytes;
          return pdf;        
        }
      } else {
        tratarRetornoErro(response.body, response.headers);
        return null;
      }
    } on Exception catch (e) {
      tratarRetornoErro(null, null, exception: e);
    }
    return null;
  }

  Future<void> visualizarPdfWeb({Filtro? filtro}) async {
    try {
      tratarFiltroRelatorio(filtro, 'TributGrupoTributario');
      visualizarRelatorioPdfWeb(filtro, 'Relatório Tribut Grupo Tributario');
    } on Exception catch (e) {
      tratarRetornoErro(null, null, exception: e);
    }
  }
  
}